Add syclSolverInverter to use oneMKL on Intel GPUs

jngkim opened the pull request:
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Add syclColverInveter and support functions in Platforms/SYCL.

## What type(s) of changes does this code introduce?
- Added QMCWaveFunctions/Fermion/syclSolverInverter.hpp and a unit test
- Modified syclBLAS and sycl_determinant_helper
- Modified CMakeLists.txt to use QMCPACK provided config file

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

- Intel GPU using nightly builds
- 
## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

ye-luo committed:
Adjust phi_vgl layout.

ye-luo committed:
Align SplineC2COMPTarget layout as SplineC2ROMPTarget

jngkim committed:
Add syclSolverInverter routines.

jngkim committed:
Merge DelayedUpdateSYCL.h

jngkim committed:
Use syclSolverInverter in DelayedUpdateSYCL.h

ye-luo committed:
Merge remote-tracking branch 'origin/develop' into sycl-allocator-solver

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

ye-luo committed:
Fix complex compilation.

ye-luo committed:
Formatting

ye-luo commented on src/Platforms/SYCL/syclBLAS.cpp:
```diff
@@ -182,6 +182,109 @@ template sycl::event gemm(sycl::queue& handle,
+  const size_t m_max         = ((m + tile_size - 1) / tile_size) * tile_size;
+  const size_t n_max         = ((n + tile_size - 1) / tile_size) * tile_size;
+
+  return q.submit([&](sycl::handler& cgh) {
```
Could you remove all handler and use the compact form of q.parallel_for with events.

ye-luo commented on src/QMCWaveFunctions/detail/SYCL/sycl_determinant_helper.hpp:
```diff
@@ -31,5 +30,14 @@ sycl::event applyW_stageV_sycl(sycl::queue& aq,
+                                   int n,
+                                   int lda,
+                                   const TMAT* a,
+                                   const INDEX* pivot,
```
Likely we should just hard-code int64_t

ye-luo replied:
I don't feel it is better than what CUDA/HIP does due to its blocking nature. CUDA code path just transfer the pivot and diagonal terms to host asynchronously.

ye-luo commented on src/QMCWaveFunctions/tests/test_syclSolverInverter.cpp:
```diff
@@ -0,0 +1,65 @@
+{
+  const int N = 911;
+
+#ifdef MIXED_PRECISION
```
Can we remove ifdef and always test both?

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Update with interop and add waits.

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

ye-luo committed:
Adopt OhmmsArray in SPOSet.

ye-luo committed:
Add data access APIs in OhmmsArray

ye-luo committed:
Renaming variables.

ye-luo committed:
Add mock-up mw_accept_rejectMove in MSD.

mdewing committed:
Add tests for Rotated SPOs using LCAO

A python script (using autograd) is used to generate reference values
for the values of the wavefunction and derivatives at one point in
space.
The test system is a helium atom with two orbitals and optionally a
Jastrow factor.

prckent committed:
Merge pull request #4059 from markdewing/test_rotated_lcao

Add tests for Rotated SPOs using LCAO

ye-luo committed:
Replace operator() with parameter packing.

ye-luo committed:
Rename data(offset) to data_at(indices)

ye-luo committed:
Replace redundant code with a function.

ye-luo committed:
Grouping doxygen comments.

ye-luo committed:
Expand a bit unit test.

ye-luo committed:
Verbose magic number

ye-luo committed:
Merge pull request #4057 from ye-luo/update-OhmmsArray

Expand OhmmsArray APIs.

ye-luo committed:
Merge branch 'develop' into adjust-layout

Hyeondeok-Shin committed:
Adjust tolerance for deterministic test for SOREP

prckent committed:
Merge pull request #4058 from ye-luo/adjust-layout

Adjust layout phi_vgl_v

prckent committed:
Merge branch 'develop' into deterministic_test

prckent committed:
Merge pull request #4060 from Hyeondeok-Shin/deterministic_test

Adjust tolerance for deterministic test for SOREP

ye-luo committed:
Fix legacy CUDA compilation issue.

Hyeondeok-Shin committed:
Adjust tolerance for unit_test_estimators for One-body densty matrices on mixed precision build

ye-luo committed:
Derive OpenMP OFFLOAD_ARCH based on HIP_ARCH.

prckent committed:
Merge pull request #4063 from ye-luo/fix-legacy-cuda

Fix legacy CUDA compilation issue on Bora

ye-luo committed:
Update OpenMP offload build instruction.

ye-luo committed:
Merge branch 'develop' into deterministic_test

ye-luo committed:
Merge pull request #4064 from Hyeondeok-Shin/deterministic_test

Adjust tolerance for unit_test_estimators for One-body density matrices on mixed precision build

prckent committed:
Merge branch 'develop' into auto-derive-arch

prckent committed:
Merge pull request #4065 from ye-luo/auto-derive-arch

Derive OpenMP OFFLOAD_ARCH based on HIP_ARCH.

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

ye-luo committed:
Update nightly_anl_bora.sh

ye-luo committed:
Update test scripts and remove unused scripts.

prckent committed:
Merge pull request #4067 from ye-luo/testing-scripts

Update bora test scripts and remove unused test scripts

mdewing committed:
Orbital rotation test with legacy driver

The gen_rotated_lcao_wf.py file is extended to generate QMC averages for
the parameter gradients.

A test for the legacy driver using one thread is introduced.

jakurzak committed:
Do not use cudaDeviceProp maxTexture1D in HIP

ye-luo committed:
Merge pull request #4070 from jakurzak/develop

Do not use cudaDeviceProp maxTexture1D in HIP

prckent committed:
Merge branch 'develop' into he_orb_rot_test

prckent committed:
Merge pull request #4069 from markdewing/he_orb_rot_test

Add orbital rotation test with legacy driver

williamfgc committed:
Add MPI support to ROCm legacy CI

For CI running on nitrogen

prckent committed:
Merge pull request #4071 from williamfgc/ci-rocm-legacy-mpi

Add MPI support to ROCm legacy CI

ye-luo committed:
Rewrite loop to avoid NVHPC hang.

ye-luo committed:
Remove omp parallel over walkers.

ye-luo committed:
Merge pull request #4073 from ye-luo/avoid-nvhpc-hang

Rewrite loop to workaround NVHPC bugs

ye-luo committed:
Merge branch 'develop' into remove-parallel

ye-luo committed:
Merge pull request #4074 from ye-luo/remove-parallel

Remove omp parallel over walkers

ye-luo committed:
Add MSD::mw_accept_rejectMove unit test

PDoakORNL committed:
drop standalone-debug for offload builds

prckent committed:
Merge pull request #4072 from ye-luo/MSD_mw_accept_reject

Add mock-up mw_accept_rejectMove in MSD.

PDoakORNL committed:
adding a message about -fstandalone-debug

ye-luo committed:
Merge branch 'develop' into remove_annoying_standalone_warning

prckent committed:
Guard div by 0

ye-luo committed:
Merge pull request #4080 from PDoakORNL/remove_annoying_standalone_warning

Drop standalone-debug for offload builds

prckent committed:
Merge branch 'develop' into fixr

prckent committed:
Merge pull request #4082 from prckent/fixr

Fix test_structure divide by 0

jngkim committed:
Merge branch 'QMCPACK:develop' into sycl-allocator-solver

jngkim committed:
Fix to avoid double-free with latest compilers.
